### DISCUSSION TITLE
My Success With Launching Shairport-Sync in Docker on Ubuntu 24.04.1

### DISCUSSION THREAD

**Discussion by DiogenesAnalytics** [mikebrady/shairport-sync#1923] [General]
Having used **Shairport-Sync** deployed with **Docker** (i.e. the `mikebrady/shairport-sync` image) for many years (almost a decade now) I had only ever used it on **Raspberry Pis** (which of course worked flawlessly). Recently I needed to deploy a *shairport-sync* instance on an Ubuntu Desktop (**24.04.1**)  installed on a *2011 iMac* and it was a bit more challenging ...

As it turns out this was in large part due to the **default output** in [ALSA](https://en.wikipedia.org/wiki/Advanced_Linux_Sound_Architecture) on Ubuntu being [Pipewire](https://ubuntuforums.org/showthread.php?t=2494723). While I was able to get **Shairport-Sync** to work on the *bare-metal host* after installing with `sudo apt install shairport-sync`, I could not get the **Docker** version to work.

But I stayed on it, and I eventually was able to discover what was the "system default" for the *audio output device* by running the following:
```
aplay -L
```

Notice that is a *capital* **L** ... which printed out:
```
null
    Discard all samples (playback) or generate zero samples (capture)
pipewire
    PipeWire Sound Server
default
    Default ALSA Output (currently PipeWire Media Server)
hw:CARD=HDMI,DEV=3
    HDA ATI HDMI, iMac
    Direct hardware device without any conversions
plughw:CARD=HDMI,DEV=3
    HDA ATI HDMI, iMac
    Hardware device with all software conversions
hdmi:CARD=HDMI,DEV=0
    HDA ATI HDMI, iMac
    HDMI Audio Output
dmix:CARD=HDMI,DEV=3
    HDA ATI HDMI, iMac
    Direct sample mixing device
hw:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    Direct hardware device without any conversions
hw:CARD=PCH,DEV=1
    HDA Intel PCH, CS4206 Digital
    Direct hardware device without any conversions
plughw:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    Hardware device with all software conversions
plughw:CARD=PCH,DEV=1
    HDA Intel PCH, CS4206 Digital
    Hardware device with all software conversions
sysdefault:CARD=PCH                                                 <---------- BINGO!!!!
    HDA Intel PCH, CS4206 Analog
    Default Audio Device
front:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    Front output / input
surround21:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    4.0 Surround output to Front and Rear speakers
surround41:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Digital
    IEC958 (S/PDIF) Digital Audio Output
dmix:CARD=PCH,DEV=0
    HDA Intel PCH, CS4206 Analog
    Direct sample mixing device
dmix:CARD=PCH,DEV=1
    HDA Intel PCH, CS4206 Digital
    Direct sample mixing device
```

Eventually, after much research and attempts to mess around with pipewire, I found a combination of **Docker options** that configured the container correctly:
```
docker run -d \
    --privileged \
    --name shairport-sync \
    --network host \
    --restart unless-stopped \
    -v /dev/snd:/dev/snd \
    mikebrady/shairport-sync \
      -v -- -d sysdefault:CARD=PCH
```

This has worked so far ... no crashing, no "sound not playing" issues ... it actually works. I hope that this info will help as many people trying to setup the *dockerized* **Shairport-Sync** on Ubuntu (specifically `24+`) as possible (because I know, as we all do, how frustrating and mind destroying solving some of these technical/configuration problems can be). 

Best of luck to you all, and I hope this helps.

P.S. 

Here is a pic of my [Owntone](https://owntone.github.io/owntone-server/) server using all 3 *dockerized* **Shairport-Sync** servers:
![Screenshot from 2024-10-22 10-10-39](https://github.com/user-attachments/assets/41aed056-4f95-4f2f-8490-2551ded3e952)

Here are the logs (sanitized) from the server running successfully:
[logs.txt](https://github.com/user-attachments/files/17477528/logs.txt)

**Comment by mikebrady** [comment]
Thanks for that useful guide!

**Reply by DiogenesAnalytics** [reply]
Just trying to give back to the community! Thanks for building this awesome tool!
